### ISSUE TITLE
Plug-ins for rendering SynBioHub pages

### ISSUE THREAD

**Issue by jakebeal** [SynBioHub/synbiohub#740]
Could we please get the ability to add modular deployment-specific plug-ins into SynBioHub pages?

For example, in the SD2 deployment of SBH, I would like to be able to add integration into the TACC environment that would not be appropriate on any other deployment.

This capability will also allow experimentation with SBH extensions and refinements without requiring updates to the core system.

We already seem to have something like this, in the form of the custom tabs for the iGEM registry entries, so I'd be hoping that either this capability exists and just needs documenting or else can be added relatively simply.

**Comment by cjmyers** [comment]
The iGEM registry entries are actually built into the codebase itself.  I'm not sure how simple it would be to add deployment-specific plug-ins.  Having code that is conditioned on, for example, config options is not difficult, but I'm not sure if you are looking for something more specific than that.

**Comment by jakebeal** [comment]
I'd like for the configuration to include "in this part of the page, call external program X to fill a page frame"

**Comment by jakebeal** [comment]
Among other things, this will allow us to experiment with displays and integrations without touching SBH or its deployment.

**Comment by zachzundel** [comment]
@cjmyers and I chatted a bit today on what is the best way to do this. 
The approach we came up with is to add the ability to add iframes to top-level component pages. In the admin panel, you could specify a URL which is called (with the share link for a given part) when a part is rendered, and the result shown in an iframe on the page. 

This has a significant drawback, which is that there is no current scheme for invalidating share links, and a malicious administrator could gain access to any parts in the repository by creating a plugin that listened for share links and saved them. Fixing this issue with share links is also a longer-term goal of ours, but for a first cut it would be a vulnerability of the plugin system.

**Comment by jakebeal** [comment]
@zachzundel This sounds like a good starting point.  How soon do you think we could get a first version of this working?  I'm hoping to put one of our folks on customer navigation and integration in the SD2 SBH instance in late December/early January.

The potential drawback, however, doesn't seem to be a big problem to me, since a malicious administrator would already have access to every part in the repository, since they're an administrator.

**Comment by zachzundel** [comment]
Unfortunately my time is a bit occupied for the next couple weeks because of finals, so I have limited cycles until about midway through December. I will have a better idea of timeframe after this weekend and will touch base again next week.

**Comment by cjmyers** [comment]
@jakebeal one clarification is that SBH does not give Admins access to all parts.  I actually need my students in my class to give me share links to see them.  If an admin does not have access to the system OR to Virtuoso, they should have no more access than a regular user.  Of course, this assumes our security scheme does not have any other holes, but it is at least designed in this way, so this would be a new vulnerability.  That being said, I would hope that you don't give admin rights to someone that you would worry about exploiting this vulnerability.

Based on my discussion with Zach, this does not sound too difficult to implement, but the plumbing may take a little time.  You can though start working on your plugin at anytime.  Assume that you are sent a GET request that includes the URL of the part.  Then, you would need to query SBH for what you need for your rendering.  This might be, for example, a <URI>/sbolnr request.  Then, generate your HTML that you are going to serve back.  In this way, we can work in parallel.

**Comment by zachzundel** [comment]
This is merged into master
